File: Homework/Mary_Antonyan/C++/30_03_2017/Chess/knight.cpp
#include "knight.hpp"
#include "board.hpp"
#include <iostream>

//Constructor
Chess::Knight::Knight(const Chess::Figure::Chessman figure, 
                        const Chess::Figure::Color color,
                        const int row, const int coloumn) 
    : Figure(figure, color, row, coloumn)
{}

//Destructor
Chess::Knight::~Knight() {}

//Method override
void Chess::Knight::step(const int row, const int coloumn) {
    if (0 <= row && 8 > row && 0 <= coloumn && 8 > coloumn) {
        if ((((row == mRow - 2 || row == mRow + 2) && 
            (coloumn == mColoumn - 1 || coloumn == mColoumn + 1)) ||
            ((row == mRow - 1 || row == mRow + 1) && 
            (coloumn == mColoumn - 2 || coloumn == mColoumn + 2))) &&
            (Chess::board[row][coloumn] == nullptr ||
             Chess::board[row][coloumn]->getColor()!= mColor)) {
            std::cout << "Move to [" << row << " , " << coloumn << "]\n";
            int tmpRow = mRow;
            int tmpCol = mColoumn;
            Chess::board[row][coloumn] = new Knight(mFigure, mColor, row, coloumn);
            Chess::board[tmpRow][tmpCol] = nullptr;
        } else {
            std::cout << "Can't move to [" << row << " , " << coloumn << "]\n";
        }
    } else {
        std::cout << "Out of board" << std::endl;
    }
}

File: Homework/Mary_Antonyan/Java/07_08_2017/Chess/src/main/java/com/itc/chess/Knight.java
package com.itc.chess;
import java.awt.Point;

/**
 * This class represents the knight(ձի) chess figure
 * @author mary
 */
public class Knight extends Figure {
    /**
     * Creates a knight object
     * 
     * @param figure figure name
     * @param color figure color
     * @param position figure current position
     */
    public Knight(Chessman figure, Color color, Point position) {
        super(figure, color, position);
    }

    /**
     * 
     * @param position new position
     * @return can the object step to new position or not
     */
    @Override
    public boolean step(Point position) {
        if (0 <= position.x && 8 > position.x && 0 <= position.y && 8 > position.y) {
            if ((((position.x == this.position.x - 2 || position.x == this.position.x + 2) && 
            (position.y == this.position.y - 1 || position.y == this.position.y + 1)) ||
            ((position.x == this.position.x - 1 || position.x == this.position.x + 1) && 
            (position.y == this.position.y - 2 || position.y == this.position.y + 2))) &&
            (Board.board[position.x][position.y] == null ||
             Board.board[position.x][position.y].getColor()!= color)) {
                System.out.println("Move to [" + position.x + " , " + position.y + "]");
                int tmpX = this.position.x;
                int tmpY = this.position.y;
                Board.board[position.x][position.y] = new Knight(figure, color, position);
                Board.board[tmpX][tmpY] = null;
                return true;
            } else {
                System.out.println("Can't move to [" + position.x + " , " + position.y + "]");
            }
        } else {
            System.out.println("Out of board");
        }
        return false;
    }
}